core/src/main/java/com/github/mizool/core/concurrent/ListenableFutureCollector.java

```diff
@@ -0,0 +1,199 @@
+    {
+        @Override
+        public void onSuccess(Void value)
+        {
+            synchronized (semaphore)
+            {
+                runningFutures.decrementAndGet();
+                if (collectionWasFinished())
+                {
+                    setResultValueIfAllFuturesFinished();
+                }
+                semaphore.notifyAll();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t)
+        {
+            synchronized (semaphore)
+            {
+                runningFutures.decrementAndGet();
+                if (collectionWasFinished())
+                {
+                    result.setException(t);
+                }
+                else
+                {
+                    throwable = t;
+                }
+                semaphore.notifyAll();
+            }
+        }
+    }
+
+    private final int maximumConcurrentFutures;
+    private final Object semaphore = new Object();
+
+    private AtomicInteger runningFutures = new AtomicInteger();
+    private Throwable throwable;
+    private SettableFuture<Void> result;
+
+    @Override
+    public Supplier<Void> supplier()
+    {
+        return () -> null;
+    }
+
+    @Override
+    public BiConsumer<Void, ListenableFuture<Void>> accumulator()
+    {
+        return (a, future) -> {
+            synchronized (semaphore)
+            {
+                verifyNoPreviousException();
+                consumeFuture(future);
+                waitWhileCapacityReached();
+            }
+        };
+    }
+
+    private void verifyNoPreviousException()
+    {
+        if (throwable != null)
+        {
+            throw new IllegalStateException("Other future raised an exception", throwable);
+        }
+    }
+
+    private void consumeFuture(ListenableFuture<Void> future)
+    {
+        runningFutures.incrementAndGet();
+        Futures.addCallback(future, new Callback());
+    }
+
+    private void waitWhileCapacityReached()
+    {
+        while (runningFutures.get() >= maximumConcurrentFutures)
+        {
+            try
+            {
+                semaphore.wait();
+            }
+            catch (InterruptedException e)
+            {
+                Thread.currentThread().interrupt();
+                throw new UncheckedInterruptedException(e);
+            }
+        }
+    }
+
+    @Override
+    public BinaryOperator<Void> combiner()
+    {
+        return (a1, a2) -> null;
+    }
+
+    @Override
+    public Function<Void, ListenableFuture<Void>> finisher()
+    {
+        return a -> {
+            synchronized (semaphore)
+            {
+                result = SettableFuture.create();
+                setResultValueIfAllFuturesFinished();
+                return result;
+            }
+        };
+    }
+
+    @Override
+    public Set<Characteristics> characteristics()
+    {
+        return EnumSet.of(Characteristics.CONCURRENT, Characteristics.UNORDERED);
```

Comment:
I now think we should not claim to be `CONCURRENT` as it gives a false impression.

[As explained here](https://stackoverflow.com/a/29920375/7641), non-`CONCURRENT` collectors can still be used from multiple threads, which would mean we can no longer guarantee that the limit is kept. But I think that's okay, provided we add a note about this in the docs.